src/hotspot/share/runtime/javaThread.hpp

```diff
@@ -497,11 +500,46 @@ class JavaThread: public Thread {
   bool pending_interrupted_exception()           { return _pending_interrupted_exception; }
   void set_pending_interrupted_exception(bool b) { _pending_interrupted_exception = b; }
 
-  bool preempting()           { return _preempt_alternate_return != nullptr; }
+  bool preempting()                              { return _preempt_alternate_return != nullptr; }
   void set_preempt_alternate_return(address val) { _preempt_alternate_return = val; }
 
-private:
+  bool at_preemptable_init()           { return _at_preemptable_init; }
+  void set_at_preemptable_init(bool b) { _at_preemptable_init = b; }
+
+#ifdef ASSERT
+  // Used for extra logging with -Xlog:continuation+preempt
+  InstanceKlass* _preempt_init_klass;
+
+  InstanceKlass* preempt_init_klass() { return _preempt_init_klass; }
+  void set_preempt_init_klass(InstanceKlass* ik) { _preempt_init_klass = ik; }
 
+  int _interp_at_preemptable_vmcall_cnt;
+  int interp_at_preemptable_vmcall_cnt() { return _interp_at_preemptable_vmcall_cnt; }
+
+  bool _interp_redoing_vm_call;
+  bool interp_redoing_vm_call() const { return _interp_redoing_vm_call; };
+
+  class AtRedoVMCall : public StackObj {
+    JavaThread* _thread;
+   public:
+    AtRedoVMCall(JavaThread* t) : _thread(t) {
+      assert(!_thread->_interp_redoing_vm_call, "");
+      _thread->_interp_redoing_vm_call = true;
+      _thread->_interp_at_preemptable_vmcall_cnt++;
+      assert(_thread->_interp_at_preemptable_vmcall_cnt > 0, "Unexpected count: %d",
+             _thread->_interp_at_preemptable_vmcall_cnt);
+    }
+    ~AtRedoVMCall() {
+      assert(_thread->_interp_redoing_vm_call, "");
+      _thread->_interp_redoing_vm_call = false;
+      _thread->_interp_at_preemptable_vmcall_cnt--;
+      assert(_thread->_interp_at_preemptable_vmcall_cnt >= 0, "Unexpected count: %d",
+             _thread->_interp_at_preemptable_vmcall_cnt);
+    }
+  };
+#endif
```

Comment:
Nit: add // ASSERT to the endif.
I always wonder if these big blocks of code added to thread.hpp and javaThread.hpp should be in some new file, and just referenced from this.  Not for this change. Just a general comment.